indico/modules/events/contributions/clone.py

```diff
@@ -98,14 +104,48 @@ def run(self, new_event, cloners, shared_data):
         self._contrib_field_map = shared_data['contribution_fields']['contrib_field_map']
         self._contrib_map = {}
         self._subcontrib_map = {}
+
         with db.session.no_autoflush:
-            self._clone_contribs(new_event)
+            if new_event == self.old_event:
+                self._clone_contrib()
+            else:
+                self._clone_contribs(new_event)
+
         self._synchronize_friendly_id(new_event)
         db.session.flush()
         return {'contrib_map': self._contrib_map, 'subcontrib_map': self._subcontrib_map}
 
-    def _clone_contribs(self, new_event):
+    def _create_new_contribution(self, event, old_contrib, excluded_attrs=None):
         attrs = (get_simple_column_attrs(Contribution) | {'own_room', 'own_venue'}) - {'abstract_id'}
+        if excluded_attrs is not None:
+            attrs -= excluded_attrs
+        new_contrib = Contribution()
+        new_contrib.populate_from_attrs(old_contrib, attrs)
+        new_contrib.subcontributions = list(self._clone_subcontribs(old_contrib.subcontributions))
+        new_contrib.acl_entries = clone_principals(ContributionPrincipal, old_contrib.acl_entries)
+        new_contrib.references = list(self._clone_references(ContributionReference, old_contrib.references))
+        new_contrib.person_links = list(self._clone_person_links(ContributionPersonLink,
+                                        old_contrib.person_links))
+        new_contrib.field_values = list(self._clone_fields(old_contrib.field_values))
+        if old_contrib.type is not None:
+            new_contrib.type = self._contrib_type_map[old_contrib.type]
+        if self.preserve_session:
+            if old_contrib.session is not None:
+                new_contrib.session = self._session_map[old_contrib.session]
+            if old_contrib.session_block is not None:
+                new_contrib.session_block = self._session_block_map[old_contrib.session_block]
+        event.contributions.append(new_contrib)
+        return new_contrib
+
+    def _clone_contrib(self):
+        new_contribution = self._create_new_contribution(self.old_event, self.contribution, {'friendly_id'})
+        if self.preserve_session:
+            entry = schedule_contribution(new_contribution, self.contribution.timetable_entry.start_dt,
+                                          session_block=new_contribution.session_block)
+            self.old_event.timetable_entries.append(entry)
```

Comment:
I really dislike the idea of using `old_event` for *any* operation that creates something new. When cloning something within the event, the new event should be the same as `old_event` anyway, but by not using the old event you remove the risk of code actually messing with the old event during normal event cloning.

Reply:
Good point.